txscript/opcode.go

```
@@ -2472,110 +2469,243 @@ func opcodeCheckMultiSig(op *parsedOpcode, vm *Engine) error {
+				return scriptError(ErrSigInvalidDataLen, "Schnorr signatures must be 64 bytes in CHECKMULTISIG")
 			}
 
 			// Parse the signature.
-			var err error
-			if vm.hasFlag(ScriptVerifyStrictEncoding) ||
-				vm.hasFlag(ScriptVerifyDERSignatures) {
+			parsedSig, err := bchec.ParseSchnorrSignature(signature)
+			if err != nil {
+				return err
+			}
 
-				parsedSig, err = bchec.ParseDERSignature(signature,
-					bchec.S256())
-			} else {
-				parsedSig, err = bchec.ParseBERSignature(signature,
-					bchec.S256())
+			if err := vm.checkPubKeyEncoding(pubKey); err != nil {
+				return err
 			}
-			sigInfo.parsed = true
+
+			// Parse the pubkey.
+			parsedPubKey, err := bchec.ParsePubKey(pubKey, bchec.S256())
 			if err != nil {
-				continue
+				return err
 			}
-			sigInfo.parsedSignature = parsedSig
-		} else {
-			// Skip to the next pubkey if the signature is invalid.
-			if sigInfo.parsedSignature == nil {
-				continue
+
+			// Generate the signature hash based on the signature hash type.
+			signatureHash, err := calcSignatureHash(script, sigHashes, hashType, &vm.tx, vm.txIdx,
+				vm.inputAmount, vm.hasFlag(ScriptVerifyBip143SigHash))
+			if err != nil {
+				vm.dstack.PushBool(false)
+				return nil
 			}
 
-			// Use the already parsed signature.
-			parsedSig = sigInfo.parsedSignature
-		}
+			var valid bool
+			if vm.sigCache != nil {
+				var sigHash chainhash.Hash
+				copy(sigHash[:], signatureHash)
 
-		if err := vm.checkPubKeyEncoding(pubKey); err != nil {
-			return err
-		}
+				valid = vm.sigCache.Exists(sigHash, parsedSig, parsedPubKey)
+				if !valid && parsedSig.Verify(signatureHash, parsedPubKey) {
+					vm.sigCache.Add(sigHash, parsedSig, parsedPubKey)
+					valid = true
+				}
+			} else {
+				valid = parsedSig.Verify(signatureHash, parsedPubKey)
+			}
 
-		// Parse the pubkey.
-		parsedPubKey, err := bchec.ParsePubKey(pubKey, bchec.S256())
-		if err != nil {
-			continue
+			if !valid {
+				success = false
+				break
```

If the signature verification fails then it should fail the script, instead of doing `break` (which might lead to returning False on stack, instead of failing the script).

I don't think the distinction would cause a consensus failure right now since NULLFAIL flag is going to be always set, but you never know, the situation could change in the future.